### ISSUE TITLE
Wasm caching

### ISSUE THREAD

**Issue by joemarshall** [pyodide/pyodide#2662]
According to https://v8.dev/blog/wasm-code-caching if you use instantiate streaming rather than instantiate, it caches the compiled webassembly in Chrome at least. Firefox is planning the feature according to an old blog.

In theory this could give massive performance improvements for minimal code changes, especially on reload.

It would need an emscripten patch which rather than calling instantiate, made a response object with a url and called streaminginstatiate on it. We'd have to see whether Chrome correctly cached things that didn't come from real fetch requests, but it might well do.

This is obviously an emscripten change but I think it would be worth patching into pyodide to provide a large test case to see if it is worth contributing upstream.

**Comment by joemarshall** [comment]
Oh hang on. You can pass a compiled module into loadWebAssemblyModule. So just needs a change to do streamingcompile in loadpackage.ts

**Comment by rth** [comment]
Are you sure it's not cached already, at least for the main module? Because there is a massive difference in load time on Chrome between the first and second load, that's not just due to caching downloads. 

> if you use instantiate streaming rather than instantiate

That post is from 2019, I can't find any relevant Chromium issues but I imagine it's possible that they also implemented caching in normal instantiate since. It would be worth checking indeed.

**Comment by hoodmane** [comment]
Hopefully they did because most of our wasm modules come out of wheels and we can't easily use instantStreaming with them

**Comment by hoodmane** [comment]
Hopefully they did because most of our wasm modules come out of wheels and we can't easily use instantStreaming with them

**Comment by ryanking13** [comment]
> Are you sure it's not cached already, at least for the main module?

Yeah I think we use (Emscripten injects) `instantiateStreaming`. It is in `pyodide.asm.js`.

```js
            function instantiateAsync() {
                if (!wasmBinary && typeof WebAssembly.instantiateStreaming === "function" && !isDataURI(wasmBinaryFile) && !isFileURI(wasmBinaryFile) && typeof fetch === "function") {
                    return fetch(wasmBinaryFile, {
                        credentials: "same-origin"
                    }).then(function(response) {
                        var result = WebAssembly.instantiateStreaming(response, info);
                        return result.then(receiveInstantiationResult, function(reason) {
                            err("wasm streaming compile failed: " + reason);
                            err("falling back to ArrayBuffer instantiation");
                            return instantiateArrayBuffer(receiveInstantiationResult)
                        })
                    })
                } else {
                    return instantiateArrayBuffer(receiveInstantiationResult)
                }
            }
```

**Comment by joemarshall** [comment]
Yeah, main module is but modules from wheels aren't. From the Api docs compilestreaming takes a response object, which can just be a wrapper for an arraybuffer, which we have anyway. As long as it has a URL and a streaming response, there's no reason it shouldn't go through the cached path. For me, if we could half the time to load numpy or scipy that would make a big difference. I'll try and have a fiddle at some point.


________________________________

**Comment by hoodmane** [comment]
I don't really understand what you are envisioning because again the wheel is a zip file and needs to be unzipped before we can instantiate the .so files so it seems like it wouldn't work. But if you have a way to handle this of course that would be great

**Comment by joemarshall** [comment]
If you have an arraybuffer you can call new Response(buffer) to create a response for it. Then some fiddling with properties to set the URL on that and you have something that should in theory be passable to compilestreaming.

The alternative would be to build a fully streaming solution using https://pypi.org/project/stream-unzip/ or similar and piping it into a stream but that would be tons of work.

**Comment by joemarshall** [comment]
Damn, it's not possible to set the URL on a response object, and the URL is the key that is used for the wasm caching and optimisation. It would be easy enough to hit the cache with a serviceworker, but that is another complexity. I might put some lib caching into a serviceworker to see how it affects performance.

**Comment by joemarshall** [comment]
If one could cache the so files from inside the wheels, then loading of so files could go concurrently with download of the wheel, with them only being reloaded if changed. Could mean that on a download, all so files could be compiling at the moment of loadpackage call, which could be quite a speedup (and enable Chrome caching and optimisation as a bonus).

**Comment by joemarshall** [comment]
I've done some messing with this now. It's non trivial, requiring a service worker and I'm not sure how much speed improvement the optimiser will bring, but in theory loading and compiling should end up faster. I'll get it tested anyway, but like all serviceworker based solutions it won't work in incognito mode (or node?). I might build a serviceworker that can get and unpack zips, plus compile  .so files, then build a minimal thing into pyodide to call it if available.


On the other hand, there are some other package loading wins that could be done without a service worker which I should check out. We'd lose out on wasm optimization of python .so modules, but it would be much simpler.

1) compile dynlibs in our code not emscripten, then pass into create module for instantiation. Then dynlib compilation can be done all at once rather than serially. This is a trivial change, although there's a bug in emscripten that requires a 1 line patch there.
2) unzip wheels with fflate not python. That way all wheels could be unpacked in parallel straight into the file system as soon as downloaded. I suspect fflate is a lot quicker than python unzip, and it supports streaming unzip, so you wouldn't have to store the wheel during download. Slightly less trivial change.

By doing these two changes, the only thing that would need to be done in any order would be the final call to module.loaddynamiclibrary. Zip decompression would run in parallel with downloads and all wasm compilation for e.g. numpy with 12 or 13 so files would run in parallel. 

I've got a package with dependencies on matplotlib, arrow and scipy, and loading it takes something like 20 seconds from cache right now, so I have an incentive to mess with this.

**Comment by joemarshall** [comment]
I messed with this quite a lot, including using a much faster unzipper coded directly in wasm, along with various experiments to make things run in workers or not. It didn't make enough difference to be worth bringing in.

 I think the main bottleneck currently is not so much the unzipping or the compiling of wasm, but the slow emscripten fs API. The whole thing would probably run way faster with the latest emscripten built with wasm_fs turned on. At that point it may be worth revisiting things like faster unzipping of wheels, but right now there's approximately a zillion JavaScript fs calls which override everything else.

**Comment by rth** [comment]
Thanks for sharing your experimentation result on this! 
Yes, I'm also looking forward to when wasm_fs is stable enough so that we can use it. Right now it looks still a bit WIP https://github.com/orgs/emscripten-core/projects/1/views/1

> I think the main bottleneck currently is not so much the unzipping or the compiling of wasm, but the slow emscripten fs API.

It probably depends on what task. For loading packages with lots of c extensions, was compilation is the bottleneck according to the profiling results https://github.com/pyodide/pyodide/issues/2288
